src/routes.rs

```diff
@@ -67,29 +68,64 @@ impl DatabaseAppState {
         terms
     }
 
-    fn courses(&self, term: Term) -> Vec<String> {
+    fn courses(&self, term: Term) -> Result<Vec<String>> {
         let Some(conn) = self.get_conn(&term) else {
-            return Vec::new();
+            return Ok(Vec::new());
         };
-        let mut stmt = conn
+
+        let courses = conn
             .prepare("SELECT subject_code, course_code FROM section")
-            .unwrap();
-
-        let mut courses = Vec::new();
-        let mut rows = stmt.query(()).unwrap();
-        while let Some(row) = rows.next().unwrap() {
-            let subject: String = row.get_unwrap("subject_code");
-            let course: String = row.get_unwrap("course_code");
-            courses.push(format!("{subject} {course}"));
-        }
-        courses
+            .context("failed to prepare courses SQL statement")?
+            .query_and_then((), |row| {
+                let subject: String = row.get("subject_code")?;
+                let course: String = row.get("course_code")?;
+                Ok(format!("{subject} {course}"))
+            })
+            .context("failed to execute query")?
+            .collect::<Result<Vec<_>>>() // quit on the first error
```

Comment:
`Result<Vec<String>>`?

Reply:
Yep. Type Inference takes care of the concrete type.